nexus/src/app/silo.rs

```diff
@@ -220,104 +219,75 @@ impl super::Nexus {
             }
         };
 
-        for group in &authenticated_subject.groups {
-            let group_name: external::Name = group.parse().map_err(|_| {
-                Error::invalid_request("Group names from an IDP must be valid")
-            })?;
+        // Gather a list of groups that the user is part of based on what the
+        // IdP sent us. Also, if the silo user provision type is Jit, create
+        // silo groups if new groups from the IdP are seen.
+
+        let mut silo_user_group_ids: Vec<Uuid> =
+            Vec::with_capacity(authenticated_subject.groups.len());
 
-            // Lookup (or create, depending on the provision type) a silo group
+        for group in &authenticated_subject.groups {
             let silo_group = match db_silo.user_provision_type {
                 db::model::UserProvisionType::Fixed => {
-                    self.silo_group_optional_lookup(
-                        opctx,
-                        &authz_silo,
-                        &group_name.clone().into(),
-                    )
-                    .await?
+                    self.db_datastore
+                        .silo_group_optional_lookup(
+                            opctx,
+                            &authz_silo,
+                            group.clone(),
+                        )
+                        .await?
                 }
 
                 db::model::UserProvisionType::Jit => {
                     let silo_group = self
                         .silo_group_lookup_or_create_by_name(
                             opctx,
                             &authz_silo,
-                            &group_name.clone().into(),
+                            &group,
                         )
                         .await?;
 
                     Some(silo_group)
                 }
             };
 
-            // and the new user to the group
             if let Some(silo_group) = silo_group {
-                self.db_datastore
-                    .silo_group_membership_create(
-                        opctx,
-                        &authz_silo,
-                        db::model::SiloGroupMembership {
-                            silo_group_id: silo_group.id(),
-                            silo_user_id: silo_user.id(),
-                        },
-                    )
-                    .await?;
+                // We're going to (potentially) modify group membership, so
+                // do an authz check here.
+                let (_authz_silo_group, ..) =
```

Comment:
I think we can skip this authz check?  We'll need an authz check in `silo_group_membership_replace_for_user()` and it seems okay to be one check for whether you can modify the user.  In practice, I assume that we're operating as Nexus itself here and it will have privileges over everybody (or else we've messed up and somebody's calling this from some other context and they shouldn't have privileges over _any_ group).

Part of my thinking here is that I think this loop will double the number of database queries and I'm not sure it's buying us anything.  We could make it a constant number of database queries by moving the `INSERT ... ON CONFLICT (external_id) DO NOTHING` into `silo_group_membership_replace_for_user()` but that's a bunch more work and probably not important now.

Reply:
changed to one lookup for user, and moved in c031ecaa